Feature: Token Budgets

chronick opened the pull request:
Adds a token budget option in LLM settings. The token budget is checked against the token count for all queries `token_budget_time_period` hours back, which is user-configurable.

Note: This is my first PR with Danswer, I've looked at the [contributing guidelines](https://github.com/danswer-ai/danswer/blob/main/CONTRIBUTING.md#-formatting-and-linting) and believe everything is formatted correctly. Let me know if that is not the case!

https://github.com/danswer-ai/danswer/issues/830

## Screenshots
### Admin Options page
![Screenshot 2024-02-12 at 6 54 29 PM](https://github.com/danswer-ai/danswer/assets/481092/b9ba7577-a8be-4505-bade-5bc13fb6660f)

### Error from backend
![Screenshot 2024-02-12 at 6 54 41 PM](https://github.com/danswer-ai/danswer/assets/481092/9e624214-b5fc-47f2-9c0c-028f246ba200)

chronick commented on backend/danswer/server/query_and_chat/token_budget.py:
```diff
@@ -0,0 +1,45 @@
+async def check_token_budget(db_session: Session = Depends(get_session)):
+    # Perform the token budget check here, possibly using `user` and `db_session` for database access if needed
+    if not is_under_token_budget(db_session):
+        raise HTTPException(status_code=429, detail="Sorry, token budget exceeded. Try again later.")
```
429 errors just show up as "No documents found" in the frontend. Is there already better way to handle this so the frontend can display a message, or do we need to build that?

Weves replied:
I will build this capability (showing the exact error detail returned) out as a follow up. For now I think this should be fine.

chronick commented on backend/danswer/server/query_and_chat/query_backend.py:
```diff
@@ -149,6 +150,7 @@ def get_answer_with_quote(
     query_request: DirectQARequest,
     user: User = Depends(current_user),
     db_session: Session = Depends(get_session),
+    _: bool = Depends(check_token_budget)
```
I've just added it to this one route for now for testing. What other routes should this be applied to?

Weves replied:
Hmm, maybe also `send-message` (in `chat_backend.py`)

chronick replied:
Done!

chronick commented on web/src/app/admin/keys/openai/page.tsx:
```diff
@@ -49,14 +57,167 @@ const ExistingKeys = () => {
+  });
+
+  const fetchConfig = async () => {
+    const response = await fetch("/api/manage/admin/token-budget-settings");
```
I've seen the `swr` library used elsewhere, is that a more idiomatic approach than this?

Weves replied:
Yea, in general we would prefer to use `swr` to standardize (although there has been some debate as to whether `swr` is the right choice 😅 )

chronick commented on web/src/app/admin/keys/openai/page.tsx:
```diff
@@ -49,14 +57,167 @@ const ExistingKeys = () => {
   );
 };
 
+const LLMOptions = () => {
```
Is there a preferred place to put this component?

Weves replied:
Here is fine! Or as a named file in the `keys/openai` dir is also pretty standard across the pages

Weves commented on backend/danswer/server/query_and_chat/token_budget.py:
```diff
@@ -0,0 +1,45 @@
+    settings_json = get_dynamic_config_store().load(TOKEN_BUDGET_SETTINGS)
+    settings = json.loads(settings_json)
+
+    is_enabled = settings.get(ENABLE_TOKEN_BUDGET, False)
```
nit: should we check this first to avoid an extra call to `get_dynamic_config_store().load` if it's disabled anyways

chronick replied:
I'm not clear on how to fix this. `is_enabled` depends on `get_dynamic_config_store()`, so I would have to call it to get this value in the first place, right? What would the alternative look like?

Weves commented:
@chronick this is awesome! Sorry for the slow turnaround here, but I added a few quick comments. I'm also happy to just address them myself in a follow up if you prefer that. 

I also noticed that the `Token Budget` section in the UI displays the default "no-token budget" settings before loading the correct settings. If possible, it would be great to show a "Loading..." / a spinner until it's ready.

chronick commented:
@Weves I've updated the PR to address your feedback and added some tests. I did have one question on one of your feedback items. Besides that, let me know if there's anything else we should do before merging!

Weves commented:
Hey @chronick, thanks for much for this change! I had to rebase to resolve some conflicts (plus I made a few styling changes to make our formatter(s) happy), but other than that looks good to me! Since I didn't have permissions to push to this branch I created a new one with the same commits (don't worry you are still the author of all lines I haven't touched as part of the rebase e.g. most of them). 

Just merged that one in—https://github.com/danswer-ai/danswer/pull/1302!

chronick commented:
Thats great, thanks for handling that! Glad to see it being used.
